crates/core/component/stake/src/component/validator_handler/uptime_tracker.rs

```diff
@@ -0,0 +1,189 @@
+    tokio::task::{AbortHandle, JoinSet},
+    tracing::{debug, error_span, instrument, trace, Instrument},
+};
+
+/// A bundle of information about a validator used to track its uptime.
+type ValidatorInformation = (IdentityKey, tendermint::PublicKey, Uptime);
+
+/// A collection of tasks retrieving [`ValidatorInformation`].
+type Lookups = JoinSet<anyhow::Result<Option<ValidatorInformation>>>;
+
+/// Tracks validator uptimes.
+///
+/// Use [`track_uptime()`] to process a block's [`CommitInfo`] and update validator uptime
+/// bookkeeping.
+///
+/// [`track_uptime()`]: Self::track_uptime
+#[async_trait]
+pub trait ValidatorUptimeTracker: StateWrite {
+    #[instrument(skip(self, last_commit_info))]
+    async fn track_uptime(&mut self, last_commit_info: &CommitInfo) -> Result<()> {
+        // Note: this probably isn't the correct height for the LastCommitInfo,
+        // which is about the *last* commit, but at least it'll be consistent,
+        // which is all we need to count signatures.
+        let height = self.get_block_height().await?;
+        let params = self.get_stake_params().await?;
+
+        // Build a mapping from addresses (20-byte truncated SHA256(pubkey)) to vote statuses.
+        let did_address_vote = last_commit_info
+            .votes
+            .as_slice()
+            .tap(|votes| {
+                if votes.is_empty() {
+                    debug!("no validators voted")
+                } else {
+                    debug!(len = %votes.len(), "collecting validator votes")
+                }
+            })
+            .into_iter()
+            .map(|vote| (vote.validator.address, vote.sig_info.is_signed()))
+            .inspect(|(address, voted)| {
+                trace!(
+                    address = %hex::encode(address),
+                    %voted,
+                    "validator vote information"
+                )
+            })
+            .collect::<BTreeMap<Address, bool>>();
+
+        // Since we don't have a lookup from "addresses" to identity keys,
+        // iterate over our app's validators, and match them up with the vote data.
+        // We can fetch all the data required for processing each validator concurrently:
+        let mut lookups = Lookups::new();
+        let mut validator_identity_stream = self.consensus_set_stream()?;
+        while let Some(identity_key) = validator_identity_stream.next().await.transpose()? {
+            self.spawn_validator_lookup_fut(identity_key, &mut lookups);
+        }
+
+        // Now process the data we fetched concurrently.
+        // Note that this will process validator uptime changes in a random order, but because they are all
+        // independent, this doesn't introduce any nondeterminism into the complete state change.
+        while let Some(data) = lookups.join_next().await.transpose()? {
+            if let Some(validator_info) = data? {
+                self.process_vote(validator_info, &did_address_vote, &params, height)
```

Comment:
I think `process_validator_uptime` is a better description of what the method actually does. Another reason to rename this is that the use of `vote` is ambiguous with ABCI++'s vote extensions (extraneous data that validators add to their consensus round messages).

Reply:
great idea. see bfb94ebd5cab23c1b77fb5322ec459705ea83370, and #4102.